clustering/common/src/main/java/org/jboss/as/clustering/controller/ServiceValueExecutor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2019, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.clustering.controller;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.jboss.msc.service.ServiceName;
+import org.wildfly.common.function.ExceptionFunction;
+
+/**
+ * Executes a given function against the captured value of a service.
+ * @author Paul Ferraro
+ */
+public class ServiceValueExecutor<T> implements ServiceValueCaptor<T>, FunctionExecutor<T> {
+
+    private final ServiceName name;
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private volatile T value = null;
+
+    public ServiceValueExecutor(ServiceName name) {
+        this.name = name;
+    }
+
+    @Override
+    public ServiceName getServiceName() {
+        return this.name;
+    }
+
+    @Override
+    public void accept(T value) {
+        Lock lock = this.lock.writeLock();
+        lock.lock();
+        try {
+            this.value = value;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public <R, E extends Exception> R execute(ExceptionFunction<T, R, E> function) throws E {
+        Lock lock = this.lock.readLock();
+        lock.lock();
+        try {
+            T value = this.value;
+            return (value != null) ? function.apply(value) : null;
```

Comment:
And above two lines will do perfect job for you.
Yes, you could miss some statistics, but this is the case anyway
because there's a race condition between retrieving the executor and invoking it.

Reply:
The primary goal of the read lock is to prevent the backing service from stopping (stopping requires a write lock) while the executor runs.  If you think the RWLock is too heavy weight (we're talking about 1 per backing service, not per metric), we could probably just use a synchronization block.  WDYT?

Reply:
You better know @pferraro what kind of behaviour of your clustering services you want to achieve.
What I wanted to point out is sometimes synchronization using intrinsic lock or volatile variable might be sufficient comparing to heavyweight RWLock.
If RWLock is your design decision here I fully accept it.

Reply:
Anyway you were right my point was there is little value to allow concurrent threads to execute(...) on a given executor.

Reply:
To be 100% clear my proposal was you might delete ReetrantRWLock and rely just on volatile variable writes and accesses. There is no need to synchronize accept(...) and execute(...) methods. As result you would have less locks involved in your statistics gathering scenarios.
The drawback would be sometimes you could miss some statistics you're not missing ATM when ReentrantRWLock is at place. The question is how acceptable such statistics loss would be for you @pferraro ?

Reply:
The concern is not about missing statistics, as you pointed out, any service starting after a volatile executor read of null would be missed, but, I think, more critically, preventing a service from stopping (i.e. blocking an invocation of ServiceValueExecutor.accept(null)) after a volatile executor read of non-null until execution completes.
I would say that if MP metrics uses a single thread for gathering metrics, then synrchonized blocks are sufficient to get the job done.  RWLocks would only be beneficial if metrics are ever gathered by a multi-threaded process.  It's worth getting this right now, in the event that this becomes a reusable pattern for delivering metrics across the application server.
@jmesnil @bstansberry  Does MP metrics use multiple WF controller threads to gather runtime metrics?

Reply:
In my opinion the answer to question whether MP metrics use single or multiple threads for statistics gathering is irrelevant because it is implementation detail that can change any time.
It is better to assume multiple threads are used.

Reply:
In any case it's single threaded.